File: internal/adapter/mongodb/user_repository.go
package mongodb

import (
	"context"
	"errors"
	"spahtmx/internal/domain"

	"log/slog"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

type UserMongoRepository struct {
	DB *mongo.Database
}

type UserMongo struct {
	ID       bson.ObjectID `bson:"_id"`
	Username string        `bson:"username"`
	Password string        `bson:"password"`
	Email    string        `bson:"email"`
	Status   bool          `bson:"status"`
}

func ToUserDomain(u UserMongo) domain.User {

	return domain.User{
		ID:       u.ID.Hex(),
		Username: u.Username,
		Password: u.Password,
		Email:    u.Email,
		Status:   u.Status,
	}

}

func FromUserDomain(user domain.User) (*UserMongo, error) {

	uid, err := bson.ObjectIDFromHex(user.ID)
	if err != nil {
		return nil, err
	}
	return &UserMongo{
		ID:       uid,
		Username: user.Username,
		Password: user.Password,
		Email:    user.Email,
		Status:   user.Status,
	}, nil
}

func (m UserMongoRepository) GetUsers(ctx context.Context) ([]domain.User, error) {
	cursor, err := m.DB.Collection("users").Find(ctx, bson.D{}, options.Find().SetLimit(10))
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var u []UserMongo
	err = cursor.All(ctx, &u)
	if err != nil {
		return nil, err
	}

	var domainUsers []domain.User
	for _, user := range u {
		domainUsers = append(domainUsers, ToUserDomain(user))
	}
	return domainUsers, nil
}

func (m UserMongoRepository) GetUser(ctx context.Context, id string) (domain.User, error) {
	objid, err := bson.ObjectIDFromHex(id)
	if err != nil {
		return domain.User{}, domain.ErrInvalidInput
	}
	user := m.DB.Collection("users").FindOne(ctx, bson.D{{Key: "_id", Value: objid}}, nil)
	var u UserMongo
	err = user.Decode(&u)
	if err != nil {
		if errors.Is(err, mongo.ErrNoDocuments) {
			return domain.User{}, domain.ErrUserNotFound
		}
		slog.Error("Database error in GetUser", "error", err, "id", id)
		return domain.User{}, domain.ErrInternal
	}
	return ToUserDomain(u), nil
}

func (m UserMongoRepository) GetByUsername(ctx context.Context, username string) (domain.User, error) {
	res := m.DB.Collection("users").FindOne(ctx, bson.D{{Key: "username", Value: username}}, nil)
	var u UserMongo
	err := res.Decode(&u)
	if err != nil {
		if errors.Is(err, mongo.ErrNoDocuments) {
			return domain.User{}, domain.ErrUserNotFound
		}
		slog.Error("Database error in GetByUsername", "error", err, "username", username)
		return domain.User{}, domain.ErrInternal
	}
	return ToUserDomain(u), nil
}

func (m UserMongoRepository) CreateUser(ctx context.Context, user domain.User) error {

	um, err := FromUserDomain(user)
	if err != nil {
		return err
	}
	_, err = m.DB.Collection("users").InsertOne(ctx, um)
	return err
}

func (m UserMongoRepository) UpdateUser(ctx context.Context, user domain.User) error {

	um, err := FromUserDomain(user)
	if err != nil {
		return err
	}
	_, err = m.DB.Collection("users").ReplaceOne(ctx, bson.D{{Key: "_id", Value: um.ID}}, um)
	return err
}

func (m UserMongoRepository) GetUserCount(ctx context.Context) string {
	return "210"
}

func (m UserMongoRepository) GetPageView(ctx context.Context) string {
	return "12345"
}

File: src/adapter/mongodb/user_repository.rs
use async_trait::async_trait;
use mongodb::{Database, bson::{doc, oid::ObjectId}, options::FindOptions};
use futures_util::stream::StreamExt;
use crate::domain::models::User;
use crate::domain::repositories::UserRepository;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct UserMongo {
    #[serde(rename = "_id")]
    pub id: ObjectId,
    pub username: String,
    pub password: String,
    pub email: String,
    pub status: bool,
}

impl From<UserMongo> for User {
    fn from(u: UserMongo) -> Self {
        User {
            id: u.id.to_hex(),
            username: u.username,
            password: u.password,
            email: u.email,
            status: u.status,
        }
    }
}

pub struct UserMongoRepository {
    pub db: Database,
}

#[async_trait]
impl UserRepository for UserMongoRepository {
    async fn get_users(&self) -> Result<Vec<User>, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<UserMongo>("users");
        let find_options = FindOptions::builder().limit(10).build();
        let mut cursor = collection.find(doc! {}).with_options(find_options).await?;
        let mut users = Vec::new();
        while let Some(user) = cursor.next().await {
            users.push(user?.into());
        }
        Ok(users)
    }

    async fn get_user(&self, id: &str) -> Result<User, Box<dyn std::error::Error + Send + Sync>> {
        let obj_id = ObjectId::parse_str(id)?;
        let collection = self.db.collection::<UserMongo>("users");
        let user = collection.find_one(doc! { "_id": obj_id }).await?;
        user.map(Into::into).ok_or_else(|| "User not found".into())
    }

    async fn get_by_username(&self, username: &str) -> Result<User, Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<UserMongo>("users");
        let user = collection.find_one(doc! { "username": username }).await?;
        user.map(Into::into).ok_or_else(|| "User not found".into())
    }

    async fn create_user(&self, user: User) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<UserMongo>("users");
        let user_mongo = UserMongo {
            id: if user.id.is_empty() { ObjectId::new() } else { ObjectId::parse_str(&user.id)? },
            username: user.username,
            password: user.password,
            email: user.email,
            status: user.status,
        };
        collection.insert_one(user_mongo).await?;
        Ok(())
    }

    async fn update_user(&self, user: User) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let collection = self.db.collection::<UserMongo>("users");
        let id = ObjectId::parse_str(&user.id)?;
        let user_mongo = UserMongo {
            id,
            username: user.username,
            password: user.password,
            email: user.email,
            status: user.status,
        };
        collection.replace_one(doc! { "_id": id }, user_mongo).await?;
        Ok(())
    }
}
